### PR TITLE
Fixes #7055: Fix reporting issue (multiple reports for one component) on branch 2.11

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://www.rudder-project.org/redmine/issues/7055

### CODE REVIEW

**Comment by VinceMacBuche** [comment]
PR updated

File: rudder-core/src/main/scala/com/normation/rudder/domain/reports/bean/ExecutionBatch.scala [hunk 1]
```diff
@@ -489,36 +485,57 @@ case class ConfigurationExecutionBatch(
...
-         val nodes = componentValues.map(value => NodeReport(value.nodeId,value.reportType,value.message))
+ def getComponentValuesRuleStatus(directiveid:DirectiveId, component:String, values:Map[Option[String], Set[String]], components:Seq[ComponentStatusReport]) : Set[ComponentValueRuleStatusReport]={
+
+   (values.flatMap{
+     case (unexpanded,values) =>
+       // Get all reports that match a value from our unexpandedValue, and group them by value
+       val componentValues = components.flatMap(_.componentValues.filter(v => values.contains(v.componentValue))).groupBy(_.componentValue)
+       for {
+         // for each value, process a reporting date
+         (value,values) <- componentValues
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/bean/ExecutionBatch.scala]: what about those that don't match ?
plus, you are using twice values in 3 lines, so when rereading, it's hard to know which one we use

File: rudder-core/src/main/scala/com/normation/rudder/domain/reports/bean/ExecutionBatch.scala [hunk 2]
```diff
@@ -489,36 +485,57 @@ case class ConfigurationExecutionBatch(
...
- def getComponentValuesRuleStatus(directiveid:DirectiveId, component:String, values:Seq[(String, Option[String])], components:Seq[ComponentStatusReport]) : Seq[ComponentValueRuleStatusReport]={
-     values.map{
-       case (value, unexpanded) =>
-         val componentValues = components.flatMap(_.componentValues.filter(_.componentValue==value))
-         val nodes = componentValues.map(value => NodeReport(value.nodeId,value.reportType,value.message))
+ def getComponentValuesRuleStatus(directiveid:DirectiveId, component:String, values:Map[Option[String], Set[String]], components:Seq[ComponentStatusReport]) : Set[ComponentValueRuleStatusReport]={
+
+   (values.flatMap{
+     case (unexpanded,values) =>
+       // Get all reports that match a value from our unexpandedValue, and group them by value
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/bean/ExecutionBatch.scala]: is this comment true ???

**Comment by VinceMacBuche** [comment]
PR updated

**Comment by VinceMacBuche** [comment]
PR updated

**Comment by VinceMacBuche** [comment]
PR updated

**Comment by fanf** [comment]
OK, so there is a lots of things strange or that I don't understand on the pr. For one, the tests are not passing. Second: it seems that we don't have values anymore, at all, in ComponentStatusReport, but only the unexpanded value. I'm pretty sure we don't want to throw it away (and only rely on the content of reports to hope for it). 
I spent several hours trying to understand what we really want, and I'm not sure of the answer now. I think that one of the reason is that our data model is not really clear, and we really miss a "report key" - what kind of unexpandedValue is used for, here, but not quite because we are still unable to correctly merge back reports on lots of edge cases. 
Redisigning the data model to assert that a we have "unexpandedvalue that can have several values that can have several messages" is a rather big change, not sure we want it to go in 3.0. 

Or we assume the fact that we are only managing unexpanded values and that we loose the value along the way, and we address the problem in 3.2. 

Ideas ?

**Comment by fanf** [comment]
Ignore the preceding comment, it was for branch 3.0 / pr #904

**Comment by fanf** [comment]
Closing this one because need a rebase, replaced by #911

**Pull request closed without merging**